ComponentKit/Core/ComponentTree/CKComponentTreeNode.mm

```diff
@@ -0,0 +1,105 @@
+
+#import "CKMutex.h"
+
+@implementation CKComponentTreeNode
+{
+  std::tuple<Class, NSUInteger> _componentKey;
+  int32_t _nodeIdentifier;
+}
+
+- (instancetype)initWithComponent:(CKComponent *)component
+                           parent:(CKComponentTreeNode *)parent
+                   previousParent:(CKComponentTreeNode *)previousParent
+                        scopeRoot:(CKComponentScopeRoot *)scopeRoot
+                     stateUpdates:(const CKComponentStateUpdateMap &)stateUpdates
+{
+
+  static int32_t nextGlobalIdentifier = 0;
+
+  if (self = [super init]) {
+
+    _nodeIdentifier = OSAtomicIncrement32(&nextGlobalIdentifier);
+    _component = component;
+
+    Class componentClass = [component class];
+    _componentKey = [parent createComponentKeyForChildWithClass:componentClass];
+
+    if (component.scopeHandle) {
+      _handle = component.scopeHandle;
+    }
+    else {
+      CKComponentTreeNode *previousNode = [previousParent childForComponentKey:_componentKey];
+      // In case we already had a component tree before.
+      if (previousNode) {
+        _handle = [previousNode.handle newHandleWithStateUpdates:stateUpdates
+                                              componentScopeRoot:scopeRoot
+                                                          parent:parent.handle];
+      }
+      // If it is the first time, we need scope handle only if there is a controller or a state.
+      else if ([componentClass controllerClass] || [componentClass initialState]) {
+        _handle = [[CKComponentScopeHandle alloc] initWithListener:scopeRoot.listener
+                                                    rootIdentifier:scopeRoot.globalIdentifier
+                                                    componentClass:componentClass
+                                               initialStateCreator:nil
+                                                            parent:parent.handle];
+      }
+
+      if (_handle) {
+        [component aquireScopeHandle:_handle];
+        [_handle resolve];
+      }
+    }
+
+    // Set the link between the parent and the child.
+    [parent setChild:self forComponentKey:_componentKey];
+    
+  }
+  return self;
+}
+
+- (id)state
+{
+  return _handle.state;
+}
+
+- (const std::tuple<Class, NSUInteger> &)componentKey
+{
+  return _componentKey;
+}
+
+- (CKComponentTreeNode *)childForComponentKey:(const std::tuple<Class, NSUInteger> &)key
+{
+  CKFatal(@"%@ shouldn't be called on a leaf component", NSStringFromSelector(_cmd));
+  return nil;
+}
+
+- (std::tuple<Class, NSUInteger>)createComponentKeyForChildWithClass:(id<CKComponentProtocol>)componentClass
+{
+  CKFatal(@"%@ shouldn't be called on a leaf component", NSStringFromSelector(_cmd));
+  return std::make_tuple(componentClass, 0);
+}
+
+- (void)setChild:(CKComponentTreeNode *)child forComponentKey:(const std::tuple<Class, NSUInteger> &)componentKey
+{
+  CKFatal(@"%@ shouldn't be called on a leaf component", NSStringFromSelector(_cmd));
```

Comment:
I'm imagining when I could see this fatal... I don't know if I'd be able to understand what I should do instead....

Reply:
This in theory could never happen, unless we have a bug in the infrastructure. I guess it's more as a safety assert for us. How would you suggest to handle such a case instead?